network/src/connectivity_manager/mod.rs

```diff
@@ -394,108 +406,171 @@ where
             })
-            .map(|(peer_id, peer)| (peer_id, &peer.addrs))
             .collect();
 
+        let max_peers_to_connect = to_connect.len();
+
+        // Sort by role for choosing
+        let mut by_role: HashMap<_, Vec<_>> = HashMap::new();
+        for (peer_id, peer) in to_connect {
+            by_role.entry(peer.role).or_default().push((peer_id, peer));
+        }
+
         // Limit the number of dialed connections from a Full Node
         // This does not limit the number of incoming connections
         // It enforces that a full node cannot have more outgoing connections than `connection_limit`
         // including in flight dials.
         let to_connect_size = if let Some(conn_limit) = self.outbound_connection_limit {
             min(
                 conn_limit.saturating_sub(self.connected.len() + self.dial_queue.len()),
-                to_connect.len(),
+                max_peers_to_connect,
             )
         } else {
-            to_connect.len()
+            max_peers_to_connect
+        };
+
+        let network_id = self.network_context.network_id();
+        let roles_to_choose = match network_id {
+            NetworkId::Validator => vec![PeerRole::Validator],
+            NetworkId::Private(_) => {
+                if network_id.is_vfn_network() {
+                    vec![PeerRole::ValidatorFullNode]
+                } else {
+                    // TODO: This is not a well defined network
+                    vec![PeerRole::PreferredUpstream, PeerRole::Upstream]
+                }
+            }
+            NetworkId::Public => vec![
+                PeerRole::PreferredUpstream,
+                PeerRole::ValidatorFullNode,
+                PeerRole::Upstream,
+            ],
         };
 
+        let mut chosen_peers: Vec<(&PeerId, &DiscoveredPeer)> = Vec::new();
+        let mut left_to_connect = to_connect_size;
+        for role in roles_to_choose {
+            if left_to_connect == 0 {
+                break;
+            }
+
+            if let Some(peers) = by_role.get(&role) {
+                // Let's continue through roles until we run out of peers to add
+                let length = peers.len();
+                if length <= left_to_connect {
+                    chosen_peers.extend_from_slice(peers);
+                    left_to_connect = left_to_connect.saturating_sub(length);
+                } else {
+                    // If we have too many to choose, choose the peers left randomly
+                    chosen_peers.extend(
+                        peers
+                            .choose_multiple(&mut self.rng, left_to_connect)
+                            .cloned(),
+                    );
+                    left_to_connect = 0;
+                }
+            }
+        }
+
+        chosen_peers
+            .iter()
+            .map(|(peer_id, peer)| (**peer_id, (*peer).clone()))
+            .collect()
+    }
```

Comment:
this function effectively does, (1) filter ineligible, (2) sort remaining by metric, (3) choose top k. I think it can be simplified a bit.

Reply:
Yeah, I've removed the sorting part for now, but it does do those 3 steps